x-pack/plugins/alerting_builtins/server/alert_types/geo_threshold/alert_type.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { schema } from '@kbn/config-schema';
+import { Service } from '../../types';
+import { BUILT_IN_ALERTS_FEATURE_ID } from '../../../common';
+import { getGeoThresholdExecutor } from './geo_threshold';
+import { AlertServices } from '../../../../alerts/server';
+
+export const GEO_THRESHOLD_ID = '.geo-threshold';
+export type TrackingEvent = 'entered' | 'exited';
+export type BoundaryType = 'entireIndex'; // Will expand to cover more, i.e. - providedShapes, etc.
+export const ActionGroupId = 'tracking threshold met';
+
+const actionVariableContextCrossingEventTimeStampLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextEnteredTimeStampLabel',
+  {
+    defaultMessage: `The time the entity's crossing was recorded to arrive at its current location`,
+  }
+);
+
+const actionVariableContextCurrentLocationLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextCurrentLocationLabel',
+  {
+    defaultMessage: 'The most recently captured location of the entity',
+  }
+);
+
+const actionVariableContextPreviousLocationLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextPreviousLocationLabel',
+  {
+    defaultMessage: 'The previously captured location of the entity',
+  }
+);
+
+const actionVariableContextCurrentBoundaryIdLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextCurrentBoundaryIdLabel',
+  {
+    defaultMessage: 'The current boundary id containing the entity (if any)',
+  }
+);
+
+const actionVariableContextPreviousBoundaryIdLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextPreviousBoundaryIdLabel',
+  {
+    defaultMessage: 'The previous boundary id containing the entity (if any)',
+  }
+);
+
+const actionVariableContextCrossingDocumentIdLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextCrossingDocumentIdLabel',
+  {
+    defaultMessage: 'The id of the crossing entity document',
+  }
+);
+
+const actionVariableContextTimeOfDetectionLabel = i18n.translate(
+  'xpack.alertingBuiltins.geoThreshold.actionVariableContextTimeOfDetectionLabel',
+  {
+    defaultMessage: 'The alert interval end time this change was recorded',
+  }
+);
+
+export function getAlertType(
+  service: Service
+): {
+  defaultActionGroupId: string;
+  executor: ({
+    previousStartedAt: currIntervalStartTime,
```

Comment:
worth noting that I believe the `previousStartedAt` will get reset to `null` when an alert is `disabled()` and then `enabled()` again - @mikecote that's still right?  Actually, I'm thinking of the alert state itself, but guessing this is implicitly part of the alert state.

For the index threshold alert, we require the user to explicitly pass a `window` of time to use, to calculate the dates used to get the range of values to test.  

There's some code here from the security solution that also uses `previousStartedAt` that you might want to look over: https://github.com/elastic/kibana/blob/4bf0932500e1d74e42bb8e4e42ec61f53fbbfd4a/x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts#L43-L50

Reply:
Agreed there's room for improvement here. Not sure if should be part of this PR or a future optimization. Will discuss!